sparkmagic/sparkmagic/livyclientlib/sparkcontroller.py

```diff
@@ -87,8 +87,8 @@ def add_session(self, name, endpoint, skip_if_exists, properties):
             return
         http_client = self._http_client(endpoint)
         session = self._livy_session(http_client, properties, self.ipython_display)
-        self.session_manager.add_session(name, session)
         session.start()
+        self.session_manager.add_session(name, session)
```

Comment:
@edwardps Can you think of any unintended side-effects of only adding the session after it's started? I agree this change makes sense, but it's been this way for many years and I don't know if there are historical reasons for the ordering

Reply:
@devstein This is a fair callout. If there is timeout issue[1], we might have session leaking. So can clean up the session for edge case caused by e.g. timeout.
Any suggestion? 

```
        try:
            session.start()
        except:
            if session.is_posted(): session.delete() # is_posted() is decided by self.status = constants.NOT_STARTED_SESSION_STATUS
            raise
        else:
            self.session_manager.add_session(name, session)
```
 
[1] - https://github.com/jupyter-incubator/sparkmagic/blob/master/sparkmagic/sparkmagic/livyclientlib/livysession.py#L151

Reply:
Hi Devin,

Made a code fix for the session leaking issue. Please let me know if you have any suggestion or comments. 
https://github.com/edwardps/sparkmagic/commit/b6c89eb03c9440acb6aa9fc9e4fea258c9da5f7b

Will update the PR after confirming the solution. 

Thanks,
Ed

Reply:
Hey Ed, thanks for the quick turnaround. The solution looks good to me. Can you add a comment to explain the logic (avoid leaking sessions)? 

Also, cc @aggFTW as he was the last to [edit this logic](https://github.com/jupyter-incubator/sparkmagic/pull/162/files#diff-5b8ca0c4c8dae6a26bd649addcc20b8ed583eb3106629668f9dfb3ac3066877dR236-R237) and [refactored it](https://github.com/jupyter-incubator/sparkmagic/pull/184/files#diff-c9a3377ce7132a4f46b4f99611416ab9cc3e1494dd3d540c2f088d28d80807b6R71). @aggFTW Do you see any issues with changing this logic?

Reply:
Sure thing. I will add more details in the comment when update this PR.

Reply:
Added info about the session leaking edge case handling in the PR description and git submit.